### PR TITLE
julia-bin: autoPatchelf manually to avoid mangling cache

### PR DESCRIPTION
**Pull request by NickCao**
## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by NickCao** [comment]
The root cause is that julia caches are also elf files, which get mangled by autoPatchelfHook, giving use the hash mismatch errors.

File: pkgs/development/compilers/julia/generic-bin.nix [hunk 1]
```diff
@@ -85,11 +85,15 @@ stdenv.mkDerivation {
   installPhase = ''
     runHook preInstall
     cp -r . $out
+  '' + lib.optionalString stdenv.hostPlatform.isLinux ''
+    autoPatchelf "$out/bin" "$out/lib" "$out/libexec"
```
Comment (thomasjm) [inline pkgs/development/compilers/julia/generic-bin.nix:89]: Maybe add a comment explaining that we're specifically avoiding `$out/share`, where the precompiled files live?

**Review by thomasjm — COMMENTED** [review]
`julia_19-bin` and `julia_110-bin` are both failing for me with the BLAS test we talked about before. Maybe we should disable this test?

I'm running the package tests now for `julia_111-bin` but they look fine so far. Thanks for figuring this out!

**Comment by NickCao** [comment]
> `julia_19-bin` and `julia_110-bin` are both failing for me with the BLAS test we talked about before. Maybe we should disable this test?

I think we shouldn't. Since it's really failing failing, it would give out wrong computations on failing platforms, which is detrimental if you rely on it to be correct.

**Review by thomasjm — APPROVED** [review]

**Comment by thomasjm** [comment]
> it would give out wrong computations on failing platforms, which is detrimental if you rely on it to be correct.

But, I can run `julia_111-bin` from `master` just fine on this machine, because it gets fetched from cache.nixos.org and the tests don't get run. Does that mean I might get wrong calculations on this machine?

**Review by Aleksanaa — APPROVED** [review]
The package is not changed on darwin so don't have to test. Others looks fine

**Comment by NickCao** [comment]
> > it would give out wrong computations on failing platforms, which is detrimental if you rely on it to be correct.
> 
> But, I can run `julia_111-bin` from `master` just fine on this machine, because it gets fetched from cache.nixos.org and the tests don't get run. Does that mean I might get wrong calculations on this machine?

`julia_111` is probably fine, since it's using a newer version of openblas with that bug fixed.

**Comment by fedeinthemix** [comment]
`julia-bin` in `nixos-24.11` is also affected. Can you back-port your fix?

**Pull request merged**